transient/sshfs.py

```diff
@@ -115,127 +108,92 @@ def provision_system(timeout: int, ssh_config: ssh.SshConfig, is_slow: bool) ->
+        tie_to_thread=False,
     )
-    client = ssh.SshClient(sshfs_config)
-    conn = client.connect_piped(timeout=connect_timeout)
-
-    try:
-        sshfs_options = "-o StrictHostKeyChecking=no -o allow_other"
-        sshfs_command = f"sudo -E sshfs {sshfs_options} {local_user}@10.0.2.2:{local_dir} {remote_dir}"
-
-        logging.info(f"Sending sshfs mount command '{sshfs_command}'")
-
-        sshfs_timeout = _SSHFS_MAX_RUN_TIME
-        if is_slow is True:
-            sshfs_timeout = _SSHFS_MAX_RUN_TIME_SLOW
-
-        # This is somewhat gnarly. The core of the issue is that sshfs is a FUSE mount,
-        # so it runs as a process (that gets backgrounded by default). SSH won't close
-        # the connection on its side until "it encounters end-of-file (eof) on the pipes
-        # connecting  to the stdout and stderr of the user program". This typically means
-        # you can do something like 'nohup <cmd> >/dev/null </dev/null 2>&1 &' to close
-        # all handles and ignore any hang ups. However, this doesn't work for SSHFS, as
-        # it spawns other processes that (I guess?) still have an open handle.
-        #
-        # This causes the SSH connetion to hang forever after the logout. Therefore, we
-        # need to close the connection on our end. So the trick here is to wait for some
-        # max time for this process to be done, then inspect the stdout for a sentinel
-        # value indicating that we _did_ get to the point where it should be OK to
-        # terminate the connection.
-        #
-        # See http://www.snailbook.com/faq/background-jobs.auto.html for some more info.
-        _, raw_stderr = conn.communicate(
-            input=f"""
-          set -e
-          sudo mkdir -p {remote_dir}
-          {sshfs_command}
-          echo TRANSIENT_SSHFS_DONE
-          exit
-        """.encode(
-                "utf-8"
-            ),
-            timeout=sshfs_timeout,
-        )
 
-        # Ensure returncode is set
-        conn.poll()
+    # Everything from here on out simply verifies that nothing went wrong.
+    if ssh_proc.stderr is not None:
+        first_stderr_line = ssh_proc.stderr.readline().decode("utf-8")
+        if "TRANSIENT_SSH_COMPLETE" not in first_stderr_line:
+            stderr = ssh_proc.communicate()[1].decode("utf-8")
+            raise RuntimeError(f"SSH connection failed with: {first_stderr_line}{stderr}")
 
-        if conn.returncode == 0:
-            # On some platforms, maybe this does actually terminate. If it does,
-            # then just return, but warn because something weird is probably happening
-            logging.warning("sshfs connection did not cause session hang as expected")
-            return
+    try:
+        ssh_proc.wait(SSHFS_MAX_RUN_TIME)
+    except subprocess.TimeoutExpired as e:
+        # This is expected.
+        pass
 
-        stderr = raw_stderr.decode("utf-8")
+    # SSHFS command failed
+    if ssh_proc.returncode:
```

Comment:
What if ssh_proc exited with 0? Doesn't the fact that ssh exited _at all_ indicate something is wrong? We would expect it to continue until we kill it, right?

Reply:
Hmm, true. At first I thought that graceful exits should be ok, but you're right. The user expects an SSHFS mount, and exiting would prevent that. I fixed this in my latest commit.